### ISSUE TITLE
Port first 2 loops in fct_ale subroutine to CUDA

### ISSUE THREAD

**Issue by goord** [ESiWACE-S1/fesom2#3]
Stepwise task list:

- [x] Create C implementation for refactored kernels
- [ ] Add preprocessor flags to switch to C or CUDA and validate
- [x] Create CUDA implementation for refactored kernels
- [ ] Create data migration routines for CUDA kernels

**Comment by isazi** [comment]
Already committed the C++ implementation of the first loop (a1).

**Comment by isazi** [comment]
Done also with the second loop (a2). Committed all code, it compiles but it's not tested.

**Comment by goord** [comment]
Fortran refactoring has been validated on cartesius

**Comment by isazi** [comment]
Should not the selection of using or not the GPU come from the Fortran side? I believe it's more flexible to have fesom2-accelerate making available both reference and accelerated versions of functions at the same time (also to avoid recompiling it for each test) and let the Fortran side pick which function to call, possibly in accordance with a user defined parameter.

**Comment by goord** [comment]
Yes the preprocessor switch should be in fesom, I checked this in yesterday evening. There is no branch for the GPU call yet though, it has yet to be inserted. And the reference code needs to be tested of course, it should be bitwise the same as the fortran code I think

**Comment by isazi** [comment]
Agreed the first step should be to test the reference code to check if 1) we understand the code correctly, and 2) if the data sharing between the two sides work. I am also writing a code generator for the first kernel, to use kernel_tuner for the tuning.

**Comment by isazi** [comment]
CUDA code for the first loop written and tested, now just need to include it in the C interface.

**Comment by isazi** [comment]
CUDA code for the second loop also written, I'll write the C interface today.
